README.md
Tokyo Edit (Android)
==================
This is a Android application for watchig "Tokyo Edit" movies.  
This application is created on Titanium Mobile.

features
---------------------------------
* in-streameplayback
* download movies
* play the download movies
* descpirtion of tokyoedit

URL
---------------------------------
<http://www.tokyoedit.com/>

CHANGELOG.md
CHANGELOG
==================

v.1.0.1
---------------------------------
* fixed when you start, an alert is displayed on some devices.

v1.0.0
---------------------------------
* first release

Resources/app/ui/my.js
(function() {
	// アプリの名前空間を分割する
	app.ui.my = {};

	// My Tokyタブのファクトリメソッド
	app.ui.my.createMyTokyoTab = function() {

		// ウインドウ生成
		var win = Titanium.UI.createWindow({
			title : 'My Tokyo Edit',
			backgroundColor : '#fff',
			exitOnClose : true,
			fullscreen : false
		});

		// マニュアルメニュー設置
		win.activity.onCreateOptionsMenu = function(e) {
			var menu = e.menu;
			var item = menu.add({
				title : 'About My Tokyo'
			});
			item.setIcon('/img/light_2x/light_info@2x.png');
			// メニュークリック時ダイアログ表示
			item.addEventListener('click', function(e) {
				var dialog = Ti.UI.createAlertDialog({
					message : 'My Tokyoではダウンロード済の動画をオフラインで視聴できます。',
					ok : 'CLOSE',
					title : 'About My Tokyo'
				}).show();
			});
		};

		// タブ作成
		var tab = Titanium.UI.createTab({
			icon : '/img/light_2x/light_download@2x.png',
			// icon: "Ti.App.Android.R.drawable.tab_icon",
			title : 'My Tokyo',
			backgroundColor : '#fff',
			window : win
		});

		// My Tokyoフォーカス時の処理
		win.addEventListener('focus', function(e) {
			// ダウンロードコンテンツや削除コンテンツを即反映させるため、表示するたびに常に動かす！

			// テーブルデータ初期化
			var movies = [];

			// ヘッダー行設定
			var headerRow = Ti.UI.createTableViewRow({
				className : 'header'
			});
			headerRow.backgroundColor = '#191919';
			headerRow.selectedBackgroundColor = '#191919';
			headerRow.height = 100;
			var headerLabel = Titanium.UI.createLabel({
				text : 'My Tokyo Edit',
				color : '#fff',
				textAlign : 'center',
				font : {
					fontSize : 32
				},
				width : 'auto',
				height : 'auto'
			});
			headerRow.add(headerLabel);
			movies.push(headerRow);

			// テーブルセクション（当面は日比谷線だけ。。。）
			var line = Ti.UI.createTableViewSection({
				headerTitle : ' ' + '東京メトロ日比谷線'
			});

			// ディレクトリファイル取得
			var dir = Titanium.Filesystem.getFile(
					Titanium.Filesystem.externalStorageDirectory, 'movies');
			if (!dir.exists()) { // ディレクトリがない場合新規作成
				dir.createDirectory();
			}

			// コンテンツ設定ファイル取得
			var fileContents = Ti.Filesystem.getFile(
					Titanium.Filesystem.resourcesDirectory,
					'/app/contents.json');
			// 文字列（JSON）へ変換
			var jsonContents = fileContents.read().toString();
			var objLine = JSON.parse(jsonContents);

			// 各テーブル行作成
			var tableFile = [];
			var tableRow = [];
			for ( var i = 0; i < objLine.hiviya.length; i++) {
				// 動画ファイル取得
				tableFile[i] = Ti.Filesystem.getFile(dir.nativePath,
						objLine.hiviya[i].file + '.mp4');
				// ダウンロード済確認
				if (tableFile[i].size > 1) {
					// ダウンロード済の場合テーブル行をセクションに追加
					tableRow[i] = app.ui.my.setTableRow(
							objLine.hiviya[i].title, objLine.hiviya[i].file
									+ '.mp4');
					line.add(tableRow[i]);
				}
			}

			// セクションをテーブルデータに挿入
			movies.push(line);
			// 最後にテーブルビュー作成！
			var tblView = Ti.UI.createTableView({
				data : movies,
				separatorColor : '#999'
			});
			win.add(tblView);

			// テーブルビュークリック処理
			tblView.addEventListener('click', function(e) {
				if (!e.row.url) {
					return false;
				} else {
					// 動画再生画面作成（ローカル再生）
					var childWin = app.ui.my.createStreamingWindow(e.row.title,
							e.row.url, tab);
					tab.open(childWin);
				}
			});

		});

		return tab;
	};

	// 動画再生画面作成ファンクション
	app.ui.my.createStreamingWindow = function(title, url, tab) {

		// ウインドウ作成
		var childWin = Titanium.UI.createWindow({
			title : title,
			backgroundColor : '#000',
			exitOnClose : false,
			fullscreen : false
		});

		// 動画プレイヤー作成
		var videoPlayer = Titanium.Media.createVideoPlayer({
			autoplay : true, // 自動再生有効
			backgroundColor : 'black',
			height : 'auto',
			width : 'auto',
			mediaControlStyle : Titanium.Media.VIDEO_CONTROL_DEFAULT,
			scalingMode : Titanium.Media.VIDEO_SCALING_ASPECT_FIT
		});

		// ディレクトリファイル取得
		var dir = Titanium.Filesystem.getFile(
				Titanium.Filesystem.externalStorageDirectory, 'movies');
		if (!dir.exists()) { // ディレクトリがない場合新規作成
			dir.createDirectory();
		}

		// 対象のローカルファイルを動画プレイヤーに追加する
		var playUrl = Ti.Filesystem.getFile(dir.nativePath, url);
		videoPlayer.url = playUrl.nativePath;
		childWin.add(videoPlayer);

		// プチ情報メニュー設置
		childWin.activity.onCreateOptionsMenu = function(e) {
			var menu = e.menu;
			var item = menu.add({
				title : '走行ルート確認'
			});
			item.setIcon('/img/light_2x/light_pin@2x.png');

			// メニュークリック時の処理
			item.addEventListener('click', function(e) {

				// ウインドウ作成
				var infoWin = Ti.UI.createWindow({
					title : title,
					backgroundColor : '#000',
					exitOnClose : false,
					fullscreen : false
				});
				// イメージビュー作成（地図画像表示）
				var arrUrl = url.split('.');
				var mapName = 'map_' + arrUrl[0] + '.jpg';
				var mapImage = Ti.UI.createImageView({
					backgroundColor : '#000',
					height : Ti.UI.FILL,
					width : Ti.UI.FILL,
					image : '/img/map/' + mapName
				});
				infoWin.add(mapImage);

				tab.open(infoWin);
			});
		};

		// 戻るボタン押下時再生終了
		childWin.addEventListener('android:back', function() {
			videoPlayer.hide();
			videoPlayer.release();
			videoPlayer = null;
			childWin.close();
		});

		return childWin;
	};

	// テーブル行作成ファンクション
	app.ui.my.setTableRow = function(title, file) {

		// ディレクトリファイル取得
		var dir = Titanium.Filesystem.getFile(
				Titanium.Filesystem.externalStorageDirectory, 'movies');
		if (!dir.exists()) { // ディレクトリがない場合新規作成
			dir.createDirectory();
		}

		// テーブル行作成
		var tableRow = Ti.UI.createTableViewRow({
			title : ' ' + title,
			backgroundColor : '#e5ebee',
			url : file,
			height : 90,
			leftImage : '/img/dark_2x/dark_play@2x.png',
			className : 'station'
		});

		// タイトルラベル作成
		var titleLabel = Titanium.UI.createLabel({
			text : title,
			color : '#1b1b1b',
			textAlign : 'left',
			left : 30,
			font : {
				fontSize : 32,
				fontWeight : 'bold'
			},
			width : 'auto',
			height : 'auto',
		});

		// 遷移ラベル作成
		var rightLabel = Titanium.UI.createLabel({
			text : ">>",
			color : '#1b1b1b',
			textAlign : 'center',
			right : 30,
			font : {
				fontSize : 32,
				fontWeight : 'bold'
			},
			width : 'auto',
			height : 'auto',
		});
		tableRow.add(titleLabel);
		tableRow.add(rightLabel);

		return tableRow;
	};

})();


Resources/app/ui/ui.js
(function() {
	// アプリの名前空間を分割する
	app.ui = {};

	/* アプリを起動するためのタブグループのファクトリメソッド */
	app.ui.createApplicationTabGroup = function() {
		// 背景色設定
		Titanium.UI.setBackgroundColor('#000');
		// タブグループ作成
		var tabGroup = Ti.UI.createTabGroup();

		var tab1 = app.ui.all.createAllTokyoTab(); // All Tokyoタブ
		var tab2 = app.ui.my.createMyTokyoTab(); // My Tokyoタブ
		var tab3 = app.ui.createTokyoEditTab(); // About Tokyo Editタブ
		// 各タブをタブグループにセット
		tabGroup.addTab(tab1);
		tabGroup.addTab(tab2);
		tabGroup.addTab(tab3);

		return tabGroup;
	};

	/* About Tokyo Editタブのファクトリメソッド */
	app.ui.createTokyoEditTab = function() {
		// ウインドウ作成
		var win = Ti.UI.createWindow({
			title : 'About Tokyo Edit',
			backgroundColor : '#000',
			exitOnClose : true,
			fullscreen : false
		});
		// タブ作成
		var tab = Ti.UI.createTab({
			icon : '/img/light_2x/light_gear@2x.png',
			title : 'Tokyo Edit',
			window : win
		});

		// WebView作成（ResourceHTML）
		var web = Ti.UI.createWebView({
			url : '/docs/examples/hero.html'
		});

		win.add(web);

		return tab;
	};

})();


Resources/app/ui/all.js
(function() {
	// アプリの名前空間を分割する
	app.ui.all = {};

	// All Tokyoタブのファクトリメソッド
	app.ui.all.createAllTokyoTab = function() {
		// ウインドウ作成
		var win = Titanium.UI.createWindow({
			title : 'All Tokyo Edit',
			backgroundColor : '#fff',
			exitOnClose : true,
			fullscreen : false
		});

		// マニュアルメニュー設置
		win.activity.onCreateOptionsMenu = function(e) {
			var menu = e.menu;
			var item = menu.add({
				title : 'About All Tokyo'
			});
			item.setIcon('/img/light_2x/light_info@2x.png');
			// メニュークリック時ダイアログ表示
			item.addEventListener('click', function(e) {
				var dialog = Ti.UI.createAlertDialog({
					message : 'All Tokyoではすべての動画をストリーミング再生で視聴できます。',
					ok : 'CLOSE',
					title : 'About All Tokyo'
				}).show();
			});
		};

		// タブ作成
		var tab = Titanium.UI.createTab({
			icon : '/img/light_2x/light_stats-bars@2x.png',
			title : 'All Tokyo',
			backgroundColor : '#fff',
			window : win
		});

		// テーブルデータ初期化
		var movies = [];

		// ヘッダー行設定
		var headerRow = Ti.UI.createTableViewRow({
			className : 'header'
		});
		headerRow.backgroundColor = '#191919';
		headerRow.selectedBackgroundColor = '#191919';
		headerRow.height = 100;
		var headerLabel = Titanium.UI.createLabel({
			text : 'All Tokyo Edit',
			color : '#fff',
			textAlign : 'center',
			font : {
				fontSize : 32
			},
			width : 'auto',
			height : 'auto'
		});
		headerRow.add(headerLabel);
		movies.push(headerRow);

		// テーブルセクション（当面は日比谷線だけ。。。）
		var line = Ti.UI.createTableViewSection({
			headerTitle : ' ' + '東京メトロ日比谷線'
		});

		// コンテンツ設定ファイル取得
		var fileContents = Ti.Filesystem.getFile(
				Titanium.Filesystem.resourcesDirectory, '/app/contents.json');
		// 文字列（JSON）へ変換
		var jsonContents = fileContents.read().toString();
		var objLine = JSON.parse(jsonContents);

		// 各テーブル行作成
		var tableRow = [];
		for ( var i = 0; i < objLine.hiviya.length; i++) {
			tableRow[i] = app.ui.all.setTableRow(objLine.hiviya[i].title,
					objLine.hiviya[i].file + '.mp4');
			line.add(tableRow[i]);
		}
		// セクションをテーブルデータに挿入
		movies.push(line);

		// 最後にテーブルビュー作成！
		var tblView = Titanium.UI.createTableView({
			data : movies,
			separatorColor : '#999'
		});
		win.add(tblView);

		// テーブルビュークリック処理
		tblView.addEventListener('click', function(e) {

			if (!e.row.url) {
				return false;
			} else {
				// 動画再生画面作成（ストリーミング再生）
				var childWin = app.ui.all.createStreamingWindow(e.row.title,
						e.row.url);
				tab.open(childWin);
			}
		});

		return tab;
	};

	// 動画再生画面作成ファンクション
	app.ui.all.createStreamingWindow = function(title, url) {

		// ウインドウ作成
		var childWin = Titanium.UI.createWindow({
			title : title,
			backgroundColor : '#000',
			exitOnClose : false,
			fullscreen : false
		});

		// 動画プレイヤー作成
		var videoPlayer = Titanium.Media.createVideoPlayer({
			autoplay : true, // 自動再生有効
			backgroundColor : 'black',
			height : 'auto',
			width : 'auto',
			mediaControlStyle : Titanium.Media.VIDEO_CONTROL_DEFAULT,
			scalingMode : Titanium.Media.VIDEO_SCALING_ASPECT_FIT
		});

		// 対象のファイル（サーバー上）を動画プレイヤーに追加する
		videoPlayer.url = url;
		childWin.add(videoPlayer);

		// ダウンロードメニュー設置
		childWin.activity.onCreateOptionsMenu = function(e) {
			var menu = e.menu;
			var item = menu.add({
				title : 'ダウンロード'
			});
			item.setIcon('/img/light_2x/light_check@2x.png');

			// メニュークリック時の処理
			item.addEventListener('click', function(e) {

				// ネットワーク環境がない場合エラーを返す（ガード節）
				if (Titanium.Network.online == false) {
					// ダイアログ生成+表示
					var dialog = Ti.UI.createAlertDialog({
						message : 'パラレルワールドに接続することが出来ません。',
						ok : 'OK',
						title : 'NETWORK ERROR'
					}).show();
				}

				// プログレスバー生成
				var ind = Titanium.UI.createProgressBar({
					width : Ti.UI.FILL,
					min : 0,
					max : 100,
					value : 0,
					height : Ti.UI.SIZE,
					color : '#888',
					message : 'ダウンロード中…',
					font : {
						fontSize : 14
					},
					top : 20
				});

				// HTTPオブジェクト生成（Ajaxと同じような記述）
				var xhr = Titanium.Network.createHTTPClient({
					// 5000ミリ秒
					timeout : 5000
				});

				// ロード完了時
				xhr.onload = function() {
					// プログレスバー削除
					childWin.remove(ind);
					// URL分解
					var arrUrl = url.split("/");
					// ファイル名（例：hiviya_01）
					var fileName = arrUrl[5];

					// ファイル保存（iOSと異なるので要注意）
					var dir = Titanium.Filesystem.getFile(
							Titanium.Filesystem.externalStorageDirectory,
							'movies');
					if (!dir.exists()) {
						dir.createDirectory();
					}
					var file = Titanium.Filesystem.getFile(dir.nativePath,
							fileName);
					var result = file.write(this.responseData);

					// 結果ダイアログ表示
					if (result) { // 成功時
						var filePath = dir + '/' + fileName;
						Ti.Media.Android.scanMediaFiles([ file.nativePath ],
								null, function(e) {
									var dialog = Ti.UI.createAlertDialog({
										message : 'ダウンロードが完了しました。',
										ok : 'OK',
										title : 'SUCCESS'
									}).show();
								});
					} else { // 失敗時
						var dialog = Ti.UI.createAlertDialog({
							message : 'ダウンロードに失敗しました。再度お試しください。',
							ok : 'OK',
							title : 'DOWNLOAD ERROR'
						}).show();
					}

					// メモリリーク対策（効果ある？）
					dir = null;
					file = null;
					this.responseData = null;
					xhr.onload = null;
					xhr.onreadystatechange = null;
					xhr.ondatastream = null;
					xhr.onerror = null;
					xhr = null;

				};

				// エラー発生時
				xhr.onerror = function(error) {
					// プログレスバー削除
					childWin.remove(ind);
					// ダイアログ生成+表示
					var dialog = Ti.UI.createAlertDialog({
						message : 'ダウンロードに失敗しました。再度お試しください。',
						ok : 'OK',
						title : 'DOWNLOAD ERROR'
					}).show();

					// メモリリーク対策
					xhr.onload = null;
					xhr.onreadystatechange = null;
					xhr.ondatastream = null;
					xhr.onerror = null;
					xhr = null;
				};

				// プログレスバー進捗率初期化
				var val = 0;
				// 通信処理時
				xhr.ondatastream = function(e) {
					// プログレスバー表示
					childWin.add(ind);
					ind.show();
					val = Math.floor(e.progress * 100);
					ind.value = val;

					// 進捗率表示（めっちゃ遅くなる）
					/*
					 * interval = setInterval(function() { //
					 * 100%の時（一瞬なのでほとんどみえない） if (e.progress == 1) { ind.value =
					 * 100; ind.message = 'ダウンロード完了'; clearInterval(interval);
					 * ind.hide(); // return; }
					 * 
					 * ind.value = val; // ind.message = 'ダウンロード中：' + val + '
					 * %'; }, 1000);
					 */
				};

				xhr.open('GET', videoPlayer.url, true);
				// リクエスト送信（引数としてJSON値を入れるとパラメータ化される）
				xhr.send();

			});
		};

		// 戻るボタン押下時再生終了
		childWin.addEventListener('android:back', function() {

			videoPlayer.hide();
			videoPlayer.release();
			videoPlayer = null;
			childWin.close();
		});

		return childWin;
	};

	// テーブル行作成ファンクション
	app.ui.all.setTableRow = function(title, file) {

		// テーブル行作成
		var tableRow = Ti.UI.createTableViewRow({
			title : ' ' + title,
			backgroundColor : '#e5ebee',
			height : 90,
			url : 'http://itosho.parallel.jp/app/tokyoedit/' + file,
			leftImage : '/img/dark_2x/dark_play@2x.png',
			className : 'station'
		});

		// タイトルラベル作成
		var titleLabel = Titanium.UI.createLabel({
			text : title,
			color : '#1b1b1b',
			textAlign : 'left',
			left : 30,
			font : {
				fontSize : 32,
				fontWeight : 'bold'
			},
			width : 'auto',
			height : 'auto',
		});
		// 遷移ラベル作成
		var rightLabel = Titanium.UI.createLabel({
			text : ">>",
			color : '#1b1b1b',
			textAlign : 'center',
			right : 30,
			font : {
				fontSize : 32,
				fontWeight : 'bold'
			},
			width : 'auto',
			height : 'auto',
		});
		tableRow.add(titleLabel);
		tableRow.add(rightLabel);

		return tableRow;
	};

})();
